README.md
# `import optoforce`

A package which simplifies connecting to and reading from optoforce sensors, using Python

This is mainly here to accompany my [blog posts](https://alknemeyer.github.io/embedded-comms-with-python/) on communicating with embedded systems using python, where the optoforce sensor is an example. I hope this package is simple enough to serve as a template for others. I see that there are [other packages](https://github.com/search?q=optoforce) on GitHub which help with this

I don't imagine it'll have many users, since [the optoforce website](https://optoforce.com) redirects to another company which doesn't even mention them, and _that_ company's [optoforce page](https://www.robotshop.com/en/optoforce.html) is pretty blank. BUT I use one, so perhaps there are others?


## Installation

```bash
python -m pip install optoforce
```


## Usage

From a python script:

```python
# 16 byte frame/single-channel 3 axis force sensor (OMD-45-FH-2000N)
from optoforce import OptoForce16 as OptoForce
from optoforce.status import no_errors

with OptoForce(speed_hz=100, filter_hz=15, zero=False) as force_sensor:
    measurement = force_sensor.read(only_latest_data=False)
    assert measurement.valid_checksum
    assert no_errors(measurement.status)

    do_stuff_with_force_readings(measurement.Fx, measurement.Fy, measurement.Fz)
```

A call to `OptoForce.read()` returns a measurement packet (a NamedTuple) containing force readings and other potentially useful data. For the specifics of each sensor model, see `Reading16`, `Reading34` and `Reading22` in [`optoforce/reading.py`](./optoforce/reading.py). For example `OptoForce34` returns force readings as  `Fx1`, `Fy1`, `Fz1`, `Fx2`, ... (not `Fx`, `Fy`, `Fz`) as there are multiple channels

It's still a little verbose, so you may want to define shortcuts for your particular application. For example, if you don't care about anything except the vertical force:

```python
with OptoForce() as force_sensor:
    read_fz = lambda: force_sensor.read(only_latest_data=False).Fz
    while True:
        print(read_fz())
```

If you haven't read the force force sensor in a little while and want to get all the packets waiting in the buffer, use:

```python
measurements = force_sensor.read_all_packets_in_buffer()
```

Or from the command line, to log to a file:

```bash
$ python -m optoforce --filename force-data.csv
```

If you want to detect and handle sensor errors as reported in the status word, look at the doc string in [`optoforce/status.py`](./optoforce/status.py). It was written for completeness after I finished my use for the OptoForce, so it hasn't been tested!


## OptoForce models supported

_Possibly_,

* 16 byte frame/single-channel 3 axis force sensor (OMD-45-FH-2000N)
* 34 byte frame/multi-channel 3 axis force sensor (4 channels)
* 22 byte frame/single-channel 6 axis force/torque sensor

but I only have access to the 16 byte frame model, so I can't test the other two. The 34 byte model [has been reported to work too](https://github.com/alknemeyer/optoforce/issues/1). I imagine this means that the 22 byte model will be fine, but the torque readings aren't scaled as I don't have that datasheet :/


## Sources

`OptoForce General DAQ - USB,CAN,UART - v1.7.pdf` was used to implement this module

The force scale parameters are from `SensitivityReport-PFH0A052.pdf`

A friend mentioned that I might not be allowed to share those docs, since the company is quite secretive, and I unfortunately haven't seen them online


## Common bugs

If you get permission errors when trying to open the serial port and you run linux, try running the code below ([source](https://stackoverflow.com/questions/27858041/oserror-errno-13-permission-denied-dev-ttyacm0-using-pyserial-from-pyth))

```bash
$ sudo chmod 666 /dev/ttyACM0  # replace with your serial port
```

Permission errors [can also happen on Windows](https://github.com/alknemeyer/optoforce/issues/1) -- you'll know that's the case when you get an error which includes `Original message: PermissionError(...)`.  That can happen when [something else is using the device](https://stackoverflow.com/questions/59993883/pyserial-permissionerror13-access-denied-none-5/63922626#63922626). In that case, making sure no other program is connected to the device should work. You could do that by unplugging the sensor, and plugging it back in again.


## Developing

See [`python-template`](https://github.com/alknemeyer/python-template/) for some general tips on Python package development.

Run unit tests using `pytest`. You should see output similar to,

```bash
$ python -m pytest
======================================== test session starts ========================================
platform linux -- Python 3.9.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/alex/dev/optoforce, configfile: pyproject.toml, testpaths: tests
collected 6 items                                                                                   

tests/test_all.py ......

========================================= 6 passed in 0.02s =========================================
```

If you have `matplotlib` installed, you can run a more visual test using

```bash
$ python test-plot.py
```

![Plot of sensor data](./test-plot.png)

### Publishing a new version

Install [flit](https://flit.readthedocs.io/en/latest/), which makes publishing packages ridiculously easy. You can install it as part of your development environment, or in a base/systemwide environment, as it's just used for publishing but isn't a requirement of the package itself.

Next, increase the `__version__` number in [`optoforce/__init__.py`](optoforce/__init__.py). Then, create a (local) tag for the commit and publish:

```bash
# make a local tag of the current commit
$ git tag v0.0.1
# push local tag to remote repo
$ git push origin v0.0.1
# generate files into dist/ and upload them to pypi
$ flit publish
```


CHANGELOG.md
# Changelog

## v0.3.0

- Fix decoder bug
- Add development documentation
- Add `valid_checksum: bool` field to readings
- Add tests and test data


tests/test_all.py
from optoforce import Reading16
import serial
import optoforce
from pathlib import Path

DATA = {
    # some longer data
    '/dev/data0': b"\xaa\x00P\x01\x00\x00\xfb\xaa\x07\x08\n-j\x00\x00\xff\xf7\xff\xfb\x00\x07\x05Q\xaa\x07\x08\n-k\x00\x00\xff\xf9\xff\xfe\x00\x0c\x05\\\xaa\x07\x08\n-l\x00\x00\xff\xf3\x00\x01\x00\n\x03Y\xaa\x07\x08\n-m\x00\x00\xff\xf3\x00\x02\x00\r\x03^\xaa\x07\x08\n-n\x00\x00\xff\xf6\x00\x03\x00\x0c\x03b\xaa\x07\x08\n-o\x00\x00\xff\xfb\xff\xff\x00\x0b\x05b\xaa\x07\x08\n-p\x00\x00\xff\xfc\xff\xfb\x00\x0b\x05`\xaa\x07\x08\n-q\x00\x00\xff\xf9\xff\xf9\x00\x0b\x05\\\xaa\x07\x08\n-r\x00\x00\xff\xf9\xff\xfb\x00\t\x05]\xaa\x07\x08\n-s\x00\x00\xff\xfd\xff\xfb\x00\t\x05b\xaa\x07\x08\n-t\x00\x00\xff\xf3\xff\xf9\x00\x0b\x05Y\xaa\x07\x08\n-u\x00\x00\x00\x00\xff\xfa\x00\x0b\x03i\xaa\x07\x08\n-v\x00\x00\xff\xfb\xff\xfc\x00\t\x05d\xaa\x07\x08\n-w\x00\x00\xff\xf9\xff\xf6\x00\n\x05^\xaa\x07\x08\n-x\x00\x00\xff\xf5\x00\x03\x00\x0c\x03k\xaa\x07\x08\n-y\x00\x00\xff\xf7\x00\x01\x00\x0b\x03k\xaa\x07\x08\n-z\x00\x00\xff\xf8\xff\xfb\x00\t\x05d\xaa\x07\x08\n",
    # the same, but shorter
    '/dev/data1': b"\xaa\x00P\x01\x00\x00\xfb\xaa\x07\x08\n-j\x00\x00\xff\xf7\xff\xfb\x00\x07\x05Q\xaa\x07\x08\n-k\x00\x00\xff\xf9\xff\xfe\x00\x0c\x05\\\xaa\x07\x08\n-l\x00\x00\xff\xf3\x00\x01\x00\n\x03Y\xaa\x07\x08\n-m\x00\x00\xff\xf3\x00",
    # a file
    'test-data.bin': (Path(__file__).parent / 'test-data.bin').read_bytes(),
}


# A dummy class, so that we don't have to read from an actual serial port
class SerialTester(serial.Serial):
    def __init__(self, *args, **kwargs):
        self._idx = 0
        self._is_open = True
        self._input_buffer_was_reset = False
        self._bytes_written = bytearray()

        for k, v in DATA.items():
            if k in args or k in kwargs.values():
                self._data = v
                return

        raise RuntimeError()

    def open(self):
        self._is_open = True

    def close(self):
        self._is_open = False

    @property
    def in_waiting(self):
        assert self._is_open
        return len(self._data) - self._idx

    @property
    def is_open(self):
        return self._is_open

    def read_until(self, expected: bytes, size=-1):
        if size > 0:
            size += self._idx
        self._idx += self._data[self._idx:size].index(expected) + len(expected)

    def read(self, size: int = 1):
        assert self._is_open
        i = self._idx
        self._idx += size
        return self._data[i: i + size]

    def write(self, data) -> int:
        self._bytes_written.extend(data)
        return len(data)

    def reset_input_buffer(self):
        self._input_buffer_was_reset = True


# monkey patch serial module
serial.Serial = SerialTester


def test_serialtester():
    s = serial.Serial(port='/dev/data0')
    assert s.is_open
    assert s.read() == b'\xaa'
    s.read_until(b'\x00\x00')
    assert s.read() == b'\xfb'
    s.close()
    assert not s.is_open


##


def test_write_correct_header():
    force_sensor = optoforce.OptoForce16(port="/dev/data0")
    force_sensor.connect()

    expected_header = bytes([170, 0, 50, 3, 10, 4, 0, 0, 237])
    assert force_sensor.opt_ser._bytes_written == expected_header  # type: ignore


EXPECTED = [
    Reading16(count=11626, status=0, Fx=-0.34069400630914826, Fy=-
              0.17359101955792156, Fz=2.3144321375433954, checksum=1361, valid_checksum=True),
    Reading16(count=11627, status=0, Fx=-0.26498422712933756, Fy=-
              0.06943640782316862, Fz=3.967597950074392, checksum=1372, valid_checksum=True),
    Reading16(count=11628, status=0, Fx=-0.4921135646687697,
              Fy=0.03471820391158431, Fz=3.3063316250619934, checksum=857, valid_checksum=True)
]


def test_read_all():
    with optoforce.OptoForce16(port="/dev/data1") as force_sensor:
        assert force_sensor.read_all_packets_in_buffer() == EXPECTED


def test_read_all_no_packets_lost():
    port = '/dev/data0'
    with optoforce.OptoForce16(port=port) as force_sensor:
        packets = force_sensor.read_all_packets_in_buffer()
        buffer = DATA[port]
        assert len(packets) == len(buffer)//16


def test_read_one():
    with optoforce.OptoForce16(port="/dev/data1") as force_sensor:
        assert force_sensor.read(only_latest_data=False) == EXPECTED[0]


def test_read_latest():
    with optoforce.OptoForce16(port="/dev/data1") as force_sensor:
        force_sensor.read(only_latest_data=True)
        assert force_sensor.opt_ser._input_buffer_was_reset  # type: ignore


optoforce/status.py
"""
Utilities to check the status word of a sensor reading, and decode any errors

The functions correspond to the status word in a packet (and are named to
correspond to the OptoForce documentation), and could be used as follows:

>>> from optoforce.status import no_errors, decode
>>> while True:
...     measurement = force_sensor.read(only_latest_data=False)
...     if no_errors(measurement.status) is False:
...         print(decode(measurement.status))

While these functions _should_ work fine, they haven't been tested on a real
device, so proceed with caution and please report back!
"""
from collections import defaultdict
from typing import List, Tuple

__all__ = ['decode', 'no_errors']

###

DAQ_TYPE = defaultdict(lambda: 'reserved', {
    0b000: 'no error',
    0b001: 'daq error',
    0b010: 'communication error'
})
SENSOR_TYPE = defaultdict(lambda: 'reserved', {
    0b000: 'no error',
    0b001: 'sensor not detected',
    0b010: 'sensor failure',
    0b100: 'temperature error'
})


def error(status: int) -> Tuple[str, str]:
    """
    >>> error((0b001 << 13) + (0b010 << 10))
    ('daq error', 'sensor failure')
    """
    return DAQ_TYPE[(status >> 13) & 0b111], SENSOR_TYPE[(status >> 10) & 0b111]  # type: ignore

###


AXES = ('Fx', 'Fy', 'Fz', 'Tx', 'Ty', 'Tz')


def overloaded_axes(status: int) -> List[str]:
    """
    >>> overloaded_axes(2**9 + 2**5)
    ['Fx', 'Ty']
    """
    return [axis for i, axis in enumerate(AXES) if (status >> (9 - i)) & 0b1 == 0b1]

###


def single_multiple(status: int):
    """
    >>> single_multiple(0b1000)
    'multiple sensors have errors'

    >>> single_multiple(0b0000)
    'only a single sensor has error (or no error)'
    """
    if status & 0b1000:
        return 'multiple sensors have errors'
    else:
        return 'only a single sensor has error (or no error)'


def sensor_number(status: int):
    """
    >>> sensor_number(0b011)
    'sensor #3'
    """
    sensornum = status & 0b111
    if sensornum == 0:
        return 'no sensor has error'
    elif sensornum <= 4:
        return f'sensor #{sensornum}'
    else:
        return 'reserved'


def multiple_sensor_selection(status: int) -> Tuple[str, str]:
    return single_multiple(status), sensor_number(status)

###


def decode(status: int):
    """
    Check the status of a frame. Returns:
    Tuple[
        Tuple[daq_status: str, sensor_status: str],
        List[overloaded_axes: str],
        Tuple[single_or_multiple: str, sensor_number: str],
    ]

    An example of a sensor gone totally wrong:

    >>> decode(0b_010_100_001_110_0_100)
    (('communication error', 'temperature error'),
     ['Fz', 'Tx', 'Ty'],
     ('only a single sensor has error (or no error)', 'sensor #4'))

    An example where everything seems fine:

    >>> check_status(0)
    (('no error', 'no error'),
     [],
     ('only a single sensor has error (or no error)', 'no sensor has error'))
    """
    return error(status), overloaded_axes(status), multiple_sensor_selection(status)


def no_errors(status: int) -> bool:
    """
    >>> no_errors(0)
    True

    >>> no_errors(0b11111111111)
    False
    """
    return status == 0

    # above is a shortcut for:

    # (daq_err, sensor_err), overloaded, (_, errs) = decode(status)
    # return (
    #     daq_err == sensor_err == 'no error'
    #     and len(overloaded) == 0
    #     and errs == 'no sensor has error'
    # )


optoforce/reading.py
from typing import NamedTuple
import struct

FX_SCALE = 1/7925 * 300   # ~= 1/26.42
FY_SCALE = 1/8641 * 300   # ~= 1/28.8
FZ_SCALE = 1/6049 * 2000  # ~= 1/3.02


class Reading16(NamedTuple):
    count: int
    status: int
    Fx: float
    Fy: float
    Fz: float
    checksum: int
    valid_checksum: bool


def read_16bytes(header: bytes, data: bytes):
    count, status, Fx, Fy, Fz, checksum = struct.unpack('>HH3hH', data)
    valid_checksum = sum(header) + sum(data[:-2]) == checksum
    return Reading16(count, status, Fx*FX_SCALE, Fy*FY_SCALE, Fz*FZ_SCALE, checksum, valid_checksum)

###


class Reading34(NamedTuple):
    count: int
    status: int
    Fx1: float
    Fy1: float
    Fz1: float
    Fx2: float
    Fy2: float
    Fz2: float
    Fx3: float
    Fy3: float
    Fz3: float
    Fx4: float
    Fy4: float
    Fz4: float
    checksum: int
    valid_checksum: bool


def read_34bytes(header: bytes, data: bytes):
    count, status, Fx1, Fy1, Fz1, Fx2, Fy2, Fz2, Fx3, Fy3, Fz3, Fx4, Fy4, Fz4, checksum = (
        struct.unpack('>HH12hH', data)
    )
    valid_checksum = sum(header) + sum(data[:-2]) == checksum
    return Reading34(count, status,
                     Fx1*FX_SCALE, Fy1*FY_SCALE, Fz1*FZ_SCALE,
                     Fx2*FX_SCALE, Fy2*FY_SCALE, Fz2*FZ_SCALE,
                     Fx3*FX_SCALE, Fy3*FY_SCALE, Fz3*FZ_SCALE,
                     Fx4*FX_SCALE, Fy4*FY_SCALE, Fz4*FZ_SCALE, checksum, valid_checksum)

###


class Reading22(NamedTuple):
    count: int
    status: int
    Fx: float
    Fy: float
    Fz: float
    Tx: int
    Ty: int
    Tz: int
    checksum: int
    valid_checksum: bool


def read_22bytes(header: bytes, data: bytes):
    count, status, Fx, Fy, Fz, Tx, Ty, Tz, checksum = (
        struct.unpack('>HH6hH', data)
    )
    valid_checksum = sum(header) + sum(data[:-2]) == checksum
    return Reading22(count, status, Fx*FX_SCALE, Fy*FY_SCALE, Fz*FZ_SCALE, Tx, Ty, Tz, checksum, valid_checksum)


optoforce/__main__.py
import logging
from datetime import datetime
import argparse
from . import OptoForce16 as OptoForce
from . import status

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description="Stream force data from an OptoForce16 to a csv file"
    )
    parser.add_argument(
        '-f', '--filename',
        default='raw-opto-log.csv',
        help='name of the file to log to (default: "raw-opto-log.csv")'
    )

    filename = parser.parse_args().filename

    logging.basicConfig(
        format='%(asctime)s | %(levelname)s | %(name)s: %(message)s',
        datefmt='%I:%M:%S',
        level=logging.INFO,
    )

    with OptoForce() as force_sensor, open(filename, 'w+') as outfile:
        outfile.write('time [H:M:S:f],Fx [N],Fy [N],Fz [N]\n')
        print(f'Writing data to {filename} - press ctrl-c to stop...')

        try:
            while True:
                # take a measurement
                m = force_sensor.read(only_latest_data=False)

                # check that the checksum is valid, and that there weren't any errors
                if not m.valid_checksum:
                    logging.warning("Got message with invalid checksum")

                    continue
                if not status.no_errors(m.status):
                    logging.warning("Got errors in the measurement status")
                    continue

                t = datetime.now().strftime('%H:%M:%S:%f')
                outfile.write(f'{t},{m.Fx},{m.Fy},{m.Fz}\n')
        except KeyboardInterrupt:
            pass


optoforce/__init__.py
"""

    optoforce

A package which simplifies connecting to and reading from optoforce sensors
"""
import serial
import logging
from serial.tools.list_ports import comports
from typing import Generic, List, Optional, TypeVar
from . import status

# typing.Literal introduced in Python v3.8
try:
    from typing import Literal  # type: ignore
except ImportError:
    from typing_extensions import Literal  # type: ignore

from .reading import Reading16, Reading22, Reading34, read_16bytes, read_22bytes, read_34bytes

__version__ = '0.3.0'
__all__ = ['OptoForce16', 'OptoForce34', 'OptoForce22', 'status']

logger = logging.getLogger(__name__)


# constants from datasheet
OPTO_PARAMS = {
    'baudrate': 1000_000,
    'stopbits': serial.STOPBITS_ONE,
    'parity': serial.PARITY_NONE,
    'bytesize': serial.EIGHTBITS,
}

SPEED_MAPPING = {
    'stop': 0,
    1000: 1,
    333: 3,
    100: 10,
    30: 33,
    10: 100,
}
# must be one of these specific values:
SPEEDS = Literal['stop', 1000, 333, 100, 30, 10]

FILTER_MAPPING = {
    'none': 0,
    500: 1,
    150: 2,
    50: 3,
    15: 4,
    5: 5,
    1.5: 6,
}
FILTERS = Literal['none', 500, 150, 50, 15, 5, 1.5]


def find_optoforce_port() -> str:
    devices = [dev for dev in comports() if dev.description == 'OptoForce DAQ']

    if len(devices) == 0:
        raise RuntimeError(f"Couldn't find an OptoForce")

    elif len(devices) == 1:
        port = devices[0].device
        assert port is not None
        return port

    else:
        raise RuntimeError(f'Found more than one OptoForce: {devices}')


T = TypeVar("T")


class _OptoForce(Generic[T]):
    # attributes which are filled in by the classes that inherit from this one
    _expected_header: bytes
    _packet_size: int
    def _decoder(self, b: bytes) -> T: ...

    def __init__(self,
                 port: Optional[str] = None,
                 speed_hz: SPEEDS = 100,
                 filter_hz: FILTERS = 15,
                 zero: bool = False):
        if speed_hz not in SPEED_MAPPING:
            raise KeyError(
                f'speed_hz must be one of: {list(SPEED_MAPPING.keys())}. Got: {speed_hz}'
            )
        if filter_hz not in FILTER_MAPPING:
            raise KeyError(
                f'filter_hz must be one of: {list(FILTER_MAPPING.keys())}. Got: {filter_hz}'
            )

        self.speed = SPEED_MAPPING[speed_hz]
        self.filter = FILTER_MAPPING[filter_hz]
        self.zero = 255 if zero else 0

        if port is None:
            self.port = find_optoforce_port()
        else:
            self.port = port

    def connect(self):
        logger.info(f'connecting at port: {self.port}')
        self.opt_ser = serial.Serial(self.port, **OPTO_PARAMS)

        # write optoforce setup code
        header = (170, 0, 50, 3)
        checksum = sum(header) + self.speed + self.filter + self.zero
        payload = (*header,
                   self.speed, self.filter, self.zero,
                   *divmod(checksum, 256))

        logger.info(f'sending configuration bytes: {payload}')
        self.opt_ser.write(payload)

    def read(self, only_latest_data: bool) -> T:
        """
        Read a packet from the serial buffer. If `only_latest_data` is True,
        and there is more than one packet waiting in the buffer, flush the buffer
        until there is only one packet left (the latest packet). Otherwise,
        just read the next packet in the buffer, even if that packet is slightly
        old.
        """
        # opt_ser.in_waiting returns the number of bytes in the buffer
        if only_latest_data and self.opt_ser.in_waiting > self._packet_size:

            # flush input to make sure we don't read old data
            self.opt_ser.reset_input_buffer()

        # Start by reading data from the input buffer until the header `expected_bytes`
        # is found. This flushes data until a packet is found
        self.opt_ser.read_until(self._expected_header)
        logger.debug('received frame header')

        # next, read the body of the packet
        raw_data = self.opt_ser.read(
            self._packet_size - len(self._expected_header)
        )

        # decode (deserialize) the bytes into regular Python data
        return self._decoder(raw_data)

    def read_all_packets_in_buffer(self) -> List[T]:
        """
        Read all packets in the buffer. Note that the `count` attribute of a packet
        can be used to tell when the packet was sent from the optoforce.
        """
        data: List[T] = []
        while self.opt_ser.in_waiting >= self._packet_size:
            data.append(self.read(only_latest_data=False))
        return data

    def close(self):
        if hasattr(self, 'opt_ser'):
            self.opt_ser.close()
            logger.info('closed connection')

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, *args):
        self.close()

    def __del__(self):
        self.close()

##


class OptoForce16(_OptoForce[Reading16]):
    _expected_header = bytes((170, 7, 8, 10))
    _packet_size = 16
    def _decoder(self, b): return read_16bytes(self._expected_header, b)


class OptoForce34(_OptoForce[Reading34]):
    _expected_header = bytes((170, 7, 8, 28))
    _packet_size = 34
    def _decoder(self, b): return read_34bytes(self._expected_header, b)


class OptoForce22(_OptoForce[Reading22]):
    _expected_header = bytes((170, 7, 8, 16))
    _packet_size = 22
    def _decoder(self, b): return read_22bytes(self._expected_header, b)

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        logger.warning("This force sensor model hasn't been tested. "
                       "Please mention on the source repo how it went! "
                       "Also, the torques aren't scaled, since I don't have that datasheet!")
